### PR TITLE
Implement lightweight Mamba wrappers

### PR DESCRIPTION
**Pull request by eric-czech**
This PR introduces many changes to make it possible to implement Caduceus models as much thinner wrappers around Mamba components.  This would make it much easier to experiment with hybrid architectures, to keep up with enhancements to Mamba, improve compute+memory efficiency and with all the extra testing here, be less likely to break.  Here is a summary of what I'm proposing to change:

### Summary

- Fork mamba (at https://github.com/Open-Athena/mamba currently) and depend on it instead of bringing much of that code in manually
  - Lightly modify it to support more extensible definitions for block, mixer, mlp and attn implementations in https://github.com/Open-Athena/mamba/commit/740fd343f5ffa488061d18c6e359eba4929c91b3
  - Add support for gradient checkpointing in https://github.com/Open-Athena/mamba/commit/a95c5c468e519011441a67954d754705843b861d
  - This implicitly addresses things like https://github.com/kuleshov-group/caduceus/issues/63
- Factor the RC-equivariance logic out of Caduceus using the idea in https://github.com/eric-czech/caduceus/pull/1
  - This means that the architecture simply processes sequences and their RC independently as different examples in a batch
  - Something else outside of the architecture, i.e. RCOperator, takes care of dealing with equivariance when its needed: https://github.com/eric-czech/caduceus/blob/c79f3c80b81406dd38c73bc9ce3d6d4a7e5c7ca5/caduceus/modeling_caduceus.py#L185-L210
- Add support for attention and MLP layers in the blocks using the new Mamba configuration (i.e. it is just passed through): https://github.com/eric-czech/caduceus/blob/c79f3c80b81406dd38c73bc9ce3d6d4a7e5c7ca5/caduceus/configuration_caduceus.py#L23-L36
- Improve testing a lot
  - Coverage is at 90% now (see "Test coverage report" below)
  - All the same RC-equivariance checks are included
  - Tests without `rcps` enabled are also included
  - RC-equivariance tests use exact equality rather than tolerances
  - Tests that the models work with both Mosaic Composer and Hugging Face Trainer in [test_training.py](https://github.com/eric-czech/caduceus/blob/c79f3c80b81406dd38c73bc9ce3d6d4a7e5c7ca5/caduceus/tests/test_training.py)
  - Tests ensure that `factory_kwargs` are used correctly by checking dtypes of all params 
  - There are tests for `CaduceusForSequenceClassification`
  - Testing fixtures and boilerplate are reduced a lot by a central [conftest.py](https://github.com/eric-czech/caduceus/blob/c79f3c80b81406dd38c73bc9ce3d6d4a7e5c7ca5/caduceus/tests/conftest.py)
  - There are hybrid architecture tests using MHA and MLP blocks in `test_caduceus_hybrid_architecture`
  - `test_automodel_roundtrip` and `test_autoconfig_roundtrip` check that HF AutoClasses work
  - There are some more debuggable, simple tests like this that build up to more complex ones with larger random matrices: https://github.com/eric-czech/caduceus/blob/c79f3c80b81406dd38c73bc9ce3d6d4a7e5c7ca5/caduceus/tests/test_modeling.py#L199-L205

### Breaking changes

There are a few things I've left out or intentionally changed:

- I removed the dependency on `pad_token_id` at https://github.com/eric-czech/caduceus/pull/4/files#diff-c3c4b88bf94f935b7c8538b62d07fb7c9f966682604e8b574bf1e5de28e675aaL480-R486
  - i.e.`loss = weighted_cross_entropy(logits, labels, loss_weights, ignore_index=self.config.pad_token_id)`
- I modified `_tokenize` in the tokenizer slightly to be more consistent with UNKs: https://github.com/eric-czech/caduceus/pull/4/commits/c79f3c80b81406dd38c73bc9ce3d6d4a7e5c7ca5#diff-8d0e3b21cdf8fcce0a9d00992fba83cc697670bd944428574cb370c567118f6aR92-R98
- I am not flipping channels and using complementary weights like in https://github.com/kuleshov-group/caduceus/blob/49e3204dfba5bd36b9c6405bca6b0396e04ab9d2/caduceus/modeling_rcps.py#L241-L245
- I do not believe that old, serialized Caduceus models or configs will work with these changes

----
<details><summary>Test coverage report</summary>

```
> pytest -s --cov=caduceus --cov-config=.coveragerc caduceus/tests/
---------- coverage: platform linux, python 3.11.11-final-0 ----------
Name                                 Stmts   Miss  Cover
--------------------------------------------------------
caduceus/__init__.py                     4      0   100%
caduceus/compat/mamba.py                26      8    69%
caduceus/configuration_caduceus.py      83      4    95%
caduceus/modeling_caduceus.py          254     29    89%
caduceus/tokenization_caduceus.py       56      3    95%
--------------------------------------------------------
TOTAL                                  423     44    90%
```

</details>

### CODE REVIEW

**Comment by eric-czech** [comment]
Noting that https://github.com/eric-czech/caduceus/pull/4/commits/54c0f6a253dda5e1f1c86fb8b94df2e15a29a08e adds `AutoConfig` and `AutoModel` tests like this, respectively:

https://github.com/eric-czech/caduceus/blob/54c0f6a253dda5e1f1c86fb8b94df2e15a29a08e/caduceus/tests/test_configuration.py#L131-L137

https://github.com/eric-czech/caduceus/blob/54c0f6a253dda5e1f1c86fb8b94df2e15a29a08e/caduceus/tests/test_modeling.py#L381-L414

<details><summary>Simple example</summary>

```python
from caduceus.modeling_caduceus import CaduceusForMaskedLM
from caduceus.configuration_caduceus import CaduceusConfig
from caduceus.tokenization_caduceus import CaduceusTokenizer
from transformers import AutoModel, AutoConfig

AutoConfig.register("caduceus", CaduceusConfig)
AutoModel.register(CaduceusConfig, CaduceusForMaskedLM)

tokenizer = CaduceusTokenizer(model_max_length=512)
config = CaduceusConfig(
    d_model=256, n_layer=8, rcps=True,
    vocab_size=tokenizer.vocab_size, 
    complement_map=tokenizer.complement_map,
    # Use MHA in layers 4 and 7
    attn_layer_idx=[3, 6],
    attn_cfg={"num_heads": 8},
    # Use MLP blocks
    d_intermediate=256,
    # You have to ensure `d_model * expand / headdim` is a multiple of 8,
    # so this needs to be modified depending on `d_model` or mamba will
    # throw an error; see:
    # https://github.com/state-spaces/mamba/issues/351#issuecomment-2167091940.
    # ssm_cfg={"headdim": 64, "expand": 2}, 
)
model = CaduceusForMaskedLM(config)
model_path = "/tmp/caduceus-model"
model.save_pretrained(model_path, safe_serialization=False)
model = AutoModel.from_pretrained(model_path)
```

</details>
